sdk/core/src/policy.rs

```diff
@@ -118,30 +68,68 @@ impl Context {
 }
 
 #[async_trait::async_trait]
-pub trait Policy: Send + Sync + std::fmt::Debug {
-    async fn send(
-        &self,
-        ctx: Context,
-        request: &mut Request,
-        next: &[Arc<dyn Policy>],
-    ) -> PolicyResult<Response>;
+pub trait Policy<R>: Send + Sync + std::fmt::Debug
+where
+    R: Send + Sync,
+{
+    async fn send(&self, ctx: Context, request: Request) -> PolicyResult<R>;
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Pipeline {
-    policies: Vec<Arc<dyn Policy>>,
+    policies: Vec<Arc<dyn Policy<Request>>>,
+    retry: Arc<dyn RetryPolicy>,
+    transport: Arc<dyn Policy<Response>>,
```

Comment:
I believe we also want to break the policies into pre-retry and post-retry policies. Can we do that now?

Reply:
Sorry I missed that. This means that we should have infallible policies that are executed only after a retry? I am not clear on this point.

Reply:
I believe there are certain policies which should *not* be redone *before* the retry policy and the retry policy only retries policies that come after it in the pipeline.

Reply:
That's correct. All our SDKs are fairly consistent on this, though JS's recent changes are a little more robust with scheduling. But to keep thing simple now, see an example at https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35.